src/sentry/testutils/pytest/mocking.py

```diff
@@ -0,0 +1,81 @@
+from __future__ import annotations
+
+from collections.abc import Callable
+from typing import ParamSpec, TypeVar
+
+# TODO: Once we're on python 3.12, we can get rid of these and change the first line of the
+# signature of `capture_return_values` to
+#   def capture_return_values[T, **P](
+P = ParamSpec("P")
+T = TypeVar("T")
+
+
+def capture_return_values(
+    fn: Callable[P, T],
+    return_values: list[T] | dict[str, list[T]],
+) -> Callable[P, T]:
+    """
+    Create a wrapped version of the given function, which stores the return value each time that
+    function is called. This is useful when you want to spy on a given function and make assertions
+    based on what it returns.
+
+    In a test, this can be used in concert with a patching context manager like so:
+
+        from unittest import mock
+        from wherever import capture_return_values
+        from animals import get_dog, get_cat
+
+        def test_getting_animals():
+            # If you're only planning to patch one function, use a list:
+
+            get_dog_return_values = []
+            wrapped_get_dog = capture_return_values(
+                get_dog, get_dog_return_values
+            )
+
+            with mock.patch(
+                "animals.get_dog", wraps=wrapped_get_dog
+            ) as get_dog_spy:
```

Comment:
This usage patter makes me wonder if it would be somehow possible to combine the two, so you'd just call:

```python
with spy("animals.get_dog", return_values=get_dog_return_values)
``` 
bringing it a bit closer to how spies work in Mocker https://pytest-mock.readthedocs.io/en/latest/usage.html#spy

But that's well beyond scope of this PR, so no worries for now.

Reply:
I was looking at this library too, I wonder if we should just add it to our dev requirements?

Reply:
Yeah, and the same way in jest you can do

```js
getDogSpy = jest.spyOn(animals, "getDog");
someFunctionThatCallsGetDog();
expect(getDogSpy).toHaveReturnedWith('maisey');
```

I'd be all for a library that makes this all easier. My way works, but it's not especially pretty. I'm going to merge for now, but would be happy to see this ripped out in favor of something better.

(Now I just gotta convince TPTB to accept [my port of `it.only`](https://github.com/getsentry/sentry/pull/46029) and we'd really be cooking...)